### ISSUE TITLE
[no-unused-vars generates] false positives for parameter properties

### ISSUE THREAD

**Issue by mkhl** [bradzacher/eslint-plugin-typescript#150]
[See the test project here](https://github.com/mkhl/eslint-plugin-typescript-issue-no-unused-vars-parameter-properties)

Code:
```typescript
class App {
	constructor(private logger: Logger) {}

	run(): void {
		this.logger.log("Hello, world!")
	}
}
```

ESLint Configuration:
```json
    "env": {
      "node": true
    },
    "plugins": [
      "typescript"
    ],
    "parser": "typescript-eslint-parser",
    "rules": {
      "no-unused-vars": "warn",
      "typescript/no-unused-vars": "warn"
    }
```

The constructor generates a property for the parameter and assign its argument to that.
Even though the parameter is used (by assigning to the property) and the property is used (by accessing it), it gets flagged as a warning.

**Comment by marchaos** [comment]
Same issue here.

**Comment by bradzacher** [comment]
The parser generates the following AST node:
https://astexplorer.net/#/gist/3ee05bfb072cdc44f13a66e89cffbb0e/e3ada19e7e961fa2a81fa8a8767e3a50cdb342be
```JSON
{
    "type": "TSParameterProperty",
    "accessibility": "private",
    "parameter": {
        "type": "Identifier",
        "name": "logger",
        "typeAnnotation": {
            "type": "TSTypeAnnotation",
            "typeAnnotation": {
                "type": "TSTypeReference",
                "typeName": {
                    "type": "Identifier",
                    "name": "Logger"
                }
            }
        }
    }
}
```
It places this node within the constructor `MethodDefinition`, not inside the `ClassBody`'s `body`.
Which is correct from an AST standpoint, but this is what causes the bug.

I'll put up a PR for it..

**Comment by bradzacher** [comment]
Question for core members (cc @nzakas / @JamesHenry / whomever else)

I've looked into it, and ESLint's `no-unused-vars` doesn't track the usage of `ClassProperty`s. It only deals with variables.

How do we want the plugin to behave in this case?
Do we just mark the `TSParameterProperty`s as used so that `no-unused-vars` ignores them entirely?

Or do we attempt to handle the `private`/`protected`/`public` use cases for them?
I started an implementation (see #161), but this only handles basic `private` usages.
ESLint gives you a lot of information on the context scope, but it might get very complex for deeply nested scopes.

**Comment by nzakas** [comment]
Sorry, I haven't looked at this project in a very long time due to health issues, so trying to get caught up.

I think the ultimate goal is to have the default ESLint `no-unused-vars` rule ignore TypeScript class properties entirely. The rule is intended to check variables and not properties, so I believe that would be the expected behavior.

How to achieve that? I'm not sure. It seems like marking the parameters as used is the quick-and-dirty way to do it, but seems like it's feeding ESLint incorrect data.

**Comment by bradzacher** [comment]
Given that tsc supports "noUnusedLocals" and "noUnusedParameters", I would err on the side of this probably being a good case to just give ESLint invalid data.

If you're okay with that approach, I'm happy to resubmit a PR with it.

**Comment by mkhl** [comment]
> It seems like marking the parameters as used is the quick-and-dirty way to do it, but seems like it's feeding ESLint incorrect data.

I’d argue that it’s not; in my view a parameter property `foo` is _both_ a parameter, accessible as `foo` in the scope of the constructor, and a property, accessible as `this.foo` in the scope of the class, and I’d argue that the definition of the latter constitutes a use of the former.
